csharp/src/IceDiscovery/Lookup.cs

```diff
@@ -162,354 +75,240 @@ public void FindAdapterById(string domainId, string adapterId, ILookupReplyPrx?
+                if (!_lookups.ContainsKey(key))
                 {
-                    return;
-                }
-
-                if (request.Retry())
-                {
-                    try
-                    {
-                        request.Invoke(_domainId, _lookups);
-                        Timer.Schedule(request, _timeout);
-                        return;
-                    }
-                    catch
-                    {
-                        // Ignore.
-                    }
+                    // Fallback: just use the given lookup reply proxy if no matching endpoint found.
+                    _lookups[key] = lookupReply;
                 }
-
-                request.Finished(null);
-                _adapterRequests.Remove(request.Id);
-                Timer.Cancel(request);
             }
+            Debug.Assert(_lookups.Count > 0);
         }
 
-        internal ValueTask<IObjectPrx?> FindAdapterAsync(string adapterId)
+        internal async ValueTask<IObjectPrx?> FindAdapterAsync(string id)
         {
+            Task<IObjectPrx?> task;
             lock (_mutex)
             {
-                if (!_adapterRequests.TryGetValue(adapterId, out AdapterRequest? request))
+                if (!_adapterReplies.TryGetValue(id, out LookupReply? replyServant))
                 {
-                    request = new AdapterRequest(this, adapterId, _retryCount);
-                    _adapterRequests.Add(adapterId, request);
+                    replyServant = new LookupReply();
+                    task = InvokeFindAdapterAsync(id, replyServant);
                 }
-
-                var task = new TaskCompletionSource<IObjectPrx?>();
-                if (request.AddCallback(task))
+                else
                 {
-                    try
-                    {
-                        request.Invoke(_domainId, _lookups);
-                        Timer.Schedule(request, _timeout);
-                    }
-                    catch
-                    {
-                        request.Finished(null);
-                        _adapterRequests.Remove(adapterId);
-                    }
+                    task = replyServant.CompletionSource.Task;
                 }
-                return new ValueTask<IObjectPrx?>(task.Task);
             }
+            return await task.ConfigureAwait(false);
         }
 
-        internal ValueTask<IObjectPrx?> FindObjectAsync(Identity id)
+        internal async Task<IObjectPrx?> InvokeFindAdapterAsync(string id, LookupReply replyServant)
         {
-            lock (_mutex)
-            {
-                if (!_objectRequests.TryGetValue(id, out ObjectRequest? request))
-                {
-                    request = new ObjectRequest(this, id, _retryCount);
-                    _objectRequests.Add(id, request);
-                }
+            // Called with mutex locked
```

Comment:
async with mutexes can lead to issues. The comment should clarify what mutex is protecting (only _adapterReplies? could we switch to a thread-safe dictionary?)

Reply:
A concurrent dictionary will avoid the need to call this method with the mutex locked, the caller still needs a mutex to ensure this is only call once, I will try to simplify this.

Reply:
I rework this a bit to avoid the need to hold the mutex while calling this